drake/solvers/MobyLCP.cpp

```diff
@@ -0,0 +1,1268 @@
+    }
+  }
+  return out;
+}
+
+template <typename Derived>
+Eigen::Index minCoeffIdx(const Eigen::MatrixBase<Derived>& in) {
+  Eigen::Index idx;
+  in.minCoeff(&idx);
+  return idx;
+}
+
+const double NEAR_ZERO = std::sqrt(std::numeric_limits<double>::epsilon());
+}
+
+namespace Drake {
+
+// Sole constructor
+MobyLCPSolver::MobyLCPSolver() : log_enabled_(false) {
+}
+
+void MobyLCPSolver::setLoggingEnabled(bool enabled) {
+  log_enabled_ = enabled;
+}
+
+std::ostream& MobyLCPSolver::LOG() const {
+  if (log_enabled_) {
+    return std::cerr;
+  }
+  return null_stream_;
+}
+
+void MobyLCPSolver::clearIndexVectors() const {
+  // clear all vectors
+  _all.clear();
+  _tlist.clear();
+  _bas.clear();
+  _nonbas.clear();
+  _j.clear();
+}
+
+bool MobyLCPSolver::solve(OptimizationProblem& prog) const {
+  // TODO ggould in principle it should be possible to render each of these as
+  // a LCP and merge the resulting problems, but I don't yet know how to do it.
+  assert(prog.getGenericConstraints().empty());
+  assert(prog.getGenericObjectives().empty());
+  assert(prog.getAllLinearConstraints().empty());
+  assert(prog.getBoundingBoxConstraints().empty());
+  assert(prog.getLinearComplementarityConstraint());
+
+  Eigen::VectorXd solution(prog.getNumVars());
+  bool solved = lcp_lemke_regularized(
+      prog.getLinearComplementarityConstraint()->getConstraint()->getM(),
+      prog.getLinearComplementarityConstraint()->getConstraint()->getq(),
+      &solution);
+  if (solved) {
+    prog.setDecisionVariableValues(solution);
+  }
+  return solved;
+}
+
+/// Fast pivoting algorithm for denerate, monotone LCPs with few nonzero,
+/// nonbasic variables
+bool MobyLCPSolver::lcp_fast(
+    const Eigen::MatrixXd& M, const Eigen::VectorXd& q, Eigen::VectorXd* z,
+    double zero_tol) const {
+  const unsigned N = q.rows();
+  const unsigned UINF = std::numeric_limits<unsigned>::max();
+
+  LOG() << "MobyLCPSolver::lcp_fast() entered" << std::endl;
+
+  // look for trivial solution
+  if (N == 0) {
+    LOG() << "MobyLCPSolver::lcp_fast() - empty problem" << std::endl;
+    z->resize(0);
+    return true;
+  }
+
+  // set zero tolerance if necessary
+  if (zero_tol < 0.0) {
+    zero_tol = M.rows() * M.lpNorm<Eigen::Infinity>()
+        * std::numeric_limits<double>::epsilon();
+  }
+
+  // prepare to setup basic and nonbasic variable indices for z
+  _nonbas.clear();
+  _bas.clear();
+
+  // see whether to warm-start
+  if (z->size() == q.size()) {
+    LOG() << "MobyLCPSolver::lcp_fast() - warm starting activated"
+        << std::endl;
+
+    for (unsigned i=0; i< z->size(); i++) {
+      if (std::fabs((*z)[i]) < zero_tol) {
+        _bas.push_back(i);
+      } else {
+        _nonbas.push_back(i);
+      }
+    }
+
+    if (log_enabled_) {
+      std::ostringstream str;
+      str << " -- non-basic indices:";
+      for (unsigned i=0; i< _nonbas.size(); i++)
+        str << " " << _nonbas[i];
+      LOG() << str.str() << std::endl;
```

Comment:
Likely not. This paradigm is an artifact of Moby's logging facility- each index would be output on a single line otherwise.

________________________________
> Date: Sun, 20 Mar 2016 13:54:35 -0700 
> From: notifications@github.com 
> To: drake@noreply.github.com 
> CC: edrumwri@hotmail.com 
> Subject: Re: [drake] Moby lcp integration (#1874) 
> 
> 
> In 
> drake/solvers/MobyLCP.cpp<https://github.com/RobotLocomotion/drake/pull/1874#discussion_r56771843>: 
> 
>> + << std::endl; 
>> + 
>> + for (unsigned i=0; i< z->size(); i++) { 
>> + if (std::fabs((*z)[i]) < zero_tol) { 
>> + _bas.push_back(i); 
>> + } else { 
>> + _nonbas.push_back(i); 
>> + } 
>> + } 
>> + 
>> + if (log_enabled_) { 
>> + std::ostringstream str; 
>> + str << " -- non-basic indices:"; 
>> + for (unsigned i=0; i< _nonbas.size(); i++) 
>> + str << " " << _nonbas[i]; 
>> + LOG() << str.str() << std::endl; 
> 
> 
> is this just faster than piping to LOG() every time? 
> 
> � 
> You are receiving this because you were mentioned. 
> Reply to this email directly or view it on 
> GitHub<https://github.com/RobotLocomotion/drake/pull/1874/files/7ba1c61c5c67d54c7abe5b17f27d04edc08002f3#r56771843>